### PR TITLE
Throw errors when doc comments are added where they're unused

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
#42617

### CODE REVIEW

File: src/test/compile-fail/issue-34222.rs [hunk 1]
```diff
@@ -13,6 +13,6 @@
 
 #[rustc_error]
 fn main() { //~ ERROR compilation successful
-    /// crash
+    // crash
     let x = 0;
 }
```
Comment (Mark-Simulacrum) [inline src/test/compile-fail/issue-34222.rs]: We might as well remove this test -- it now tests nothing.

**Review by Mark-Simulacrum — COMMENTED** [review]
I feel like this needs a warning cycle... or at least a crater run.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum: I remember that we talked about it but can't remember where and when so couldn't find the discussion... But putting warnings instead is fine for me.

**Comment by nikomatsakis** [comment]
Crater run initiated.

- [ ] Building toolchains [1](https://tools.taskcluster.net/groups/KZ-iBjRrTb2PYuVSfeTwgQ/tasks/KZ-iBjRrTb2PYuVSfeTwgQ/runs/0/logs/public%2Flogs%2Flive.log), [2](https://tools.taskcluster.net/groups/aOAQCkxTRmaM4-SBfTJuYw/tasks/aOAQCkxTRmaM4-SBfTJuYw/runs/0/logs/public%2Flogs%2Flive.log)
- [ ] Running tests
- [ ] Await quiescence
- [ ] Generate report

**Comment by nikomatsakis** [comment]
I was still unable to build one of the toolchains (toolchain 1, if you follow the links above):

```
error[E0585](B: found a documentation comment that doesn't document anything(B
   (B--> (B/root/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-bidi-0.3.3/src/lib.rs:656:9(B
    (B|(B
656(B (B| (B        /// BidiTest:69635 (AL ET EN)(B
    (B| (B        (B^^^^^^^^^^^^^^^^^^^^^^^^^^^^^(B
    (B|(B
    (B= (Bhelp(B: doc comments must come before what they document, maybe a comment was intended with `//`?(B
error[E0585](B: found a documentation comment that doesn't document anything(B
   (B--> (B/root/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-bidi-0.3.3/src/lib.rs:798:9(B
    (B|(B
798(B (B| (B        /// BidiTest:946 (LRI PDI)(B
    (B| (B        (B^^^^^^^^^^^^^^^^^^^^^^^^^^(B

```

**Comment by GuillaumeGomez** [comment]
So I need to make a lint out of this. Looking into https://github.com/rust-lang/rust/blob/master/src/librustc_lint/builtin.rs#L725

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -2426,6 +2425,12 @@ impl<'a> Parser<'a> {
             expr.map(|mut expr| {
                 attrs.extend::<Vec<_>>(expr.attrs.into());
                 expr.attrs = attrs;
+                if if let Some(ref doc) = expr.attrs.iter().find(|x| x.is_sugared_doc) {
+                    self.span_fatal_err(doc.span, Error::UselessDocComment).emit();
+                    true
+                } else { false } {
+                    return expr;
+                }
```
Comment (kennytm) [inline src/libsyntax/parse/parser.rs]: 😨

Why not simply 

```rust
if let Some(ref doc) = ... {
    self.span_fatal_err(...).emit();
    return expr;
}
```

Comment (GuillaumeGomez) [inline src/libsyntax/parse/parser.rs]: Tired I was.

**Comment by nikomatsakis** [comment]
@GuillaumeGomez I can't easily set settings on a crater run, unfortunately --- though making it a lint (usually with deny-by-default) is actually the right way to do these tests (which I had forgotten).

**Comment by nikomatsakis** [comment]
Errors from travis:

```
[00:55:54] error[E0560]: struct `ast::Block` has no field named `attrs`

[00:55:54]    --> /checkout/src/libsyntax/parse/mod.rs:907:41

[00:55:54]     |

[00:55:54] 907 |                                         attrs: ThinVec::new(),

[00:55:54]     |                                         ^^^^^^ `ast::Block` does not have this field

[00:55:54] 
```

**Comment by GuillaumeGomez** [comment]
@nikomatsakis: oups, I removed your commits by mistake. :-/ Can you put it back please?

**Comment by nikomatsakis** [comment]
Crater run initiated.

- [ ] Building compilers [old](https://tools.taskcluster.net/groups/rThL3eNCR2SSQpTg1fbK9w/tasks/rThL3eNCR2SSQpTg1fbK9w/details), [new](https://tools.taskcluster.net/groups/-NxXEUcsRXCrOy4eraozBg/tasks/-NxXEUcsRXCrOy4eraozBg/details)

**Comment by nikomatsakis** [comment]
Huh, we encountered an error, I guess because one of our primary targets makes use of this feature. Seems to be actually a use in error-chain...

**Comment by nikomatsakis** [comment]
See [failure here](https://tools.taskcluster.net/groups/-NxXEUcsRXCrOy4eraozBg/tasks/-NxXEUcsRXCrOy4eraozBg/runs/0/logs/public%2Flogs%2Flive.log).

**Comment by alexcrichton** [comment]
I think @nikomatsakis is on vacation now, so changing to 

r? @nrc

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -723,6 +723,45 @@ impl EarlyLintPass for IllegalFloatLiteralPattern {
 }
 
 declare_lint! {
+    pub UNUSED_DOC_COMMENT,
+    Deny,
+    "detects use of useless doc comments"
```
Comment (nrc) [inline src/librustc_lint/builtin.rs]: I would not say "useless", rather that they will be ignored by rustdoc (theoretically, another tool could use them).

File: src/librustc_lint/builtin.rs [hunk 4]
```diff
@@ -723,6 +723,45 @@ impl EarlyLintPass for IllegalFloatLiteralPattern {
...
+        lint_array![UNUSED_DOC_COMMENT]
+    }
+}
+
+impl UnusedDocComment {
+    fn warn_if_doc<'a, 'tcx,
+                   I: Iterator<Item=&'a ast::Attribute>,
+                   C: LintContext<'tcx>>(&self, mut attrs: I, cx: &C) {
+        if let Some(attr) = attrs.find(|a| a.is_value_str() && a.check_name("doc")) {
+            cx.struct_span_lint(UNUSED_DOC_COMMENT, attr.span, "unused doc comment").emit();
```
Comment (nrc) [inline src/librustc_lint/builtin.rs]: again, I'd prefer to be explicit that unused means unused by rustdoc

File: src/librustc_lint/builtin.rs [hunk 5]
```diff
@@ -724,7 +724,7 @@ impl EarlyLintPass for IllegalFloatLiteralPattern {
 
 declare_lint! {
     pub UNUSED_DOC_COMMENT,
-    Warn,
+    Deny,
```
Comment (nrc) [inline src/librustc_lint/builtin.rs]: Should be `Warn` now

**Review by nrc — APPROVED** [review]
Looks good. I had a couple of nits about the phrasing, plus simulacrum's comment could be addressed. Otherwise r+ (iiuc, crater showed an issue, but if this is just a warning lint, that is OK. If I'm wrong, we'll want to fix that first).

File: src/librustc_lint/builtin.rs (hunk 5 above)
Comment (GuillaumeGomez) [inline src/librustc_lint/builtin.rs]: I'll just remove @nikomatsakis's commits.

**Comment by GuillaumeGomez** [comment]
@nrc: Updated and applied your comments.

File: src/test/compile-fail/useless_comment.rs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![deny(unused_doc_comment)]
+
+fn foo() {
+    /// a //~ ERROR unused doc comment
```
Comment (nrc) [inline src/test/compile-fail/useless_comment.rs]: I think this test needs updating now

File: src/test/compile-fail/issue-34222.rs (hunk 1 above)
Comment (nrc) [inline src/test/compile-fail/issue-34222.rs]: And this test still needs removing

**Comment by nrc** [comment]
r=me with the last two comments addressed (and tests passing)

**Comment by GuillaumeGomez** [comment]
Removed the useless test and updated the new one. Then:

@bors: r=nrc

**Comment by GuillaumeGomez** [comment]
There is an issue on the macro doc comments apparently. Taking a look.

@bors: r-

**Comment by GuillaumeGomez** [comment]
With all the external issues fixed, let's give it another try!

@bors: r=nrc

**Comment by kennytm** [comment]
dist stage2-rls failed. Legit. `cargo` changed the signature of `Packages::from_flags` but `rls` did not catch up.

```
[01:19:29] error[E0061]: this function takes 4 parameters but 3 parameters were supplied
[01:19:29]    --> /checkout/src/tools/rls/src/build/cargo.rs:108:37
[01:19:29]     |
[01:19:29] 108 |     let spec = Packages::from_flags(opts.all, &opts.exclude, &opts.package)
[01:19:29]     |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 4 parameters
[01:19:29] 
[01:19:30] error: aborting due to previous error
[01:19:30] 
[01:19:30] error: Could not compile `rls`.
```

(Travis link, because bors is dead or something: https://travis-ci.org/rust-lang/rust/builds/258881061)

**Comment by GuillaumeGomez** [comment]
And let's retry!

@bors: r=nrc

**Comment by kennytm** [comment]
@GuillaumeGomez Err no, the current `rls` is [still using the old signature](https://github.com/rust-lang-nursery/rls/blob/56be5b3bd5af5d7d21a0804003b2fc902921fd74/src/build/cargo.rs#L108). You need to file a PR for `rls` to fix that first.

**Comment by GuillaumeGomez** [comment]
Ah indeed! Thanks!

@bors: r-

**Comment by GuillaumeGomez** [comment]
Ok, so it has been updated to new cargo version. Let's give it another try! :)

@bors: r=nrc

**Comment by kennytm** [comment]
@bors retry p=-2 

(trying to get bors running properly)

**Pull request merged**